feat(deployment): add basic Deployment resource

skyscrapr opened the pull request:
## Summary

Adds Deployment resource with basic fields. More complete support for Deployments will come in future PRs (see https://github.com/PrefectHQ/terraform-provider-prefect/issues/238).

Related to https://github.com/PrefectHQ/terraform-provider-prefect/issues/238

Depends on https://github.com/PrefectHQ/terraform-provider-prefect/pull/201.

## Notes

- API docs: https://app.prefect.cloud/api/docs#tag/Deployments
- This PR is already big, so we're implementing some core fields here and [the rest](https://github.com/PrefectHQ/terraform-provider-prefect/issues/238) will be implemented in future PR(s)
- CI `Acceptance tests` are sometimes failing with `503 service unavailable` (and sometimes [passing](https://github.com/PrefectHQ/terraform-provider-prefect/actions/runs/10379690891/job/28738722369)), but I tested the newly-added test by hand and it passed:

   ```
   $ go test ./... -count=1 -v -run TestAccResource_deployment
   ?       github.com/prefecthq/terraform-provider-prefect [no test files]
   .. etc ...
   ?       github.com/prefecthq/terraform-provider-prefect/internal/utils  [no test files]
   testing: warning: no tests to run
   PASS
   ok      github.com/prefecthq/terraform-provider-prefect/internal/provider/datasources   0.339s [no tests to run]
   === RUN   TestAccResource_deployment
   === PAUSE TestAccResource_deployment
   === CONT  TestAccResource_deployment
   --- PASS: TestAccResource_deployment (3.66s)
   PASS
   ok      github.com/prefecthq/terraform-provider-prefect/internal/provider/resources     4.252s
   ```

## To do

- [x] Run a basic test to create and destroy a Flow + Deployment
- [x] Implement necessary attributes (others can come in a separate PR)
- [x] Ensure current acceptance tests pass
- [x] Expand acceptance tests as necessary

skyscrapr committed:
added main scaffolding to support deployments

skyscrapr committed:
added main scaffolding to support deployments

skyscrapr committed:
add support for flows

skyscrapr committed:
add support for flows

skyscrapr committed:
fixup flows

skyscrapr committed:
fixup flows

skyscrapr committed:
fixup deployments

skyscrapr committed:
fixup deployments

skyscrapr committed:
add support for deployment access

skyscrapr committed:
add support for deployment access

skyscrapr committed:
set required attributes

skyscrapr committed:
set required attributes

skyscrapr committed:
add support for deployment access

skyscrapr committed:
add support for deployment access

skyscrapr committed:
fix flow import

skyscrapr committed:
fix flow import

skyscrapr committed:
fix imports for workspace resources

skyscrapr committed:
fix imports for workspace resources

skyscrapr committed:
Merge branch 'deployment_access' of https://github.com/skyscrapr/terraform-provider-prefect into deployment_access

skyscrapr committed:
fixup tests to avoid hardcoding

skyscrapr committed:
tidy up

skyscrapr committed:
remove deployments

skyscrapr committed:
remove deployments

skyscrapr commented:
Deployments are dependent on flows
You cannot create a deployment without a flow_id. 

I split flows into #201. But I'm unsure. How to create this PR if it requires another PR to be merged first. The branch will not pass tests. 

If you explain how to do this, happy to update the PR and remove flows.

I have a third PR for deployment access that I haven't opened since it is depending on deployments and only makes the problem worse.

skyscrapr commented:
The other option is to merge #201 and that should remove flows from #202 automatically.

mitchnielsen commented:
Yeah, I was thinking we could set the Deployment PR's branch target to this Flow PR, but I don't think there's a way to do that with forks. 

Let's mark the Deployment PR as draft for now, and when this PR merges we'll revisit that one. That way, any changes we make here don't need to also be applied to your Deployment PR.

skyscrapr committed:
updates from PR review

skyscrapr committed:
revert change to test name prefix

skyscrapr committed:
updated flow resource to force recreation on change of tags

skyscrapr committed:
Update internal/client/flows.go

Co-authored-by: Mitchell Nielsen <mitchnielsen@users.noreply.github.com>

skyscrapr committed:
Update internal/provider/resources/flow.go

Co-authored-by: Mitchell Nielsen <mitchnielsen@users.noreply.github.com>

skyscrapr committed:
change model to state as per PR review

skyscrapr committed:
Update internal/client/flows.go

Co-authored-by: Mitchell Nielsen <mitchnielsen@users.noreply.github.com>

skyscrapr committed:
Update internal/provider/resources/flow.go

Co-authored-by: Mitchell Nielsen <mitchnielsen@users.noreply.github.com>

skyscrapr committed:
Update internal/provider/resources/flow_test.go

Co-authored-by: Mitchell Nielsen <mitchnielsen@users.noreply.github.com>

skyscrapr committed:
Update internal/provider/resources/flow.go

Co-authored-by: Mitchell Nielsen <mitchnielsen@users.noreply.github.com>

skyscrapr committed:
change model variable to plan as per PR review

skyscrapr committed:
tidyup import logic

skyscrapr committed:
add prefect_flow examples

skyscrapr committed:
add prefect_flow examples

skyscrapr committed:
modify import logic as suggested in PR review

skyscrapr committed:
modify import logic as suggested in PR review

skyscrapr committed:
update docs for flow import

skyscrapr committed:
Update internal/provider/resources/flow.go

Co-authored-by: Edward Park <ed.sh.park@gmail.com>

skyscrapr committed:
Update internal/provider/resources/flow.go

Co-authored-by: Edward Park <ed.sh.park@gmail.com>

skyscrapr committed:
add support for deployment resource

skyscrapr committed:
tidy up some varible names and comments

skyscrapr committed:
regenerate docs

skyscrapr committed:
fixup import for deployments

skyscrapr committed:
add deployment resource examples

skyscrapr committed:
fixup typo in flow resrource example

skyscrapr committed:
regenerate docs

skyscrapr committed:
tidy up some code

mitchnielsen commented:
Not sure why the [Terraform docs job](https://github.com/PrefectHQ/terraform-provider-prefect/actions/runs/9637652715/job/26957379060?pr=202) is failing.

tehioant commented:
@skyscrapr do you need help finishing this pr ? I would love to see this feature.
If I can help in any way 🙂

mitchnielsen commented:
Hi @skyscrapr, thanks again for submitting this. I'm catching up from PTO from a move 📦 

I'd like to take a look at this one this week. Could you please address the conflicts?

mitchnielsen commented:
@tehioant - thank you for reaching out and offering your help 🤝 Once @skyscrapr resolves the conflicts I'll dive in and test out the changes. If we need any extra hands I'll reach out!

parkedwards commented:
hey @skyscrapr - if you're open to it, we'd be be happy to help help take this PR home + squash any of the remaining merge conflicts.  we can either push to your forked repo on this branch, or re-create a branch with these changes.  let us know

skyscrapr commented:
@parkedwards  yes please go ahead. This PR contains basic functionality and still needs to be extended to support schedules etc. However it would be good to get this merged so that we can work on adding support for schedules etc. I'm currently tied up in my day job. :-) But can help support later this week.

skyscrapr commented:
Best to push to my forked branch please as this will allow us to continue working on deployment access and schedules.

parkedwards commented:
@skyscrapr totally makes sense. we'll push to this branch. thank you for all your work on this so far - between me, you and @mitchnielsen we'll be able to steer this home

mitchnielsen committed:
Merge branch 'main' of github.com:PrefectHQ/terraform-provider-prefect into resource_deployment

mitchnielsen committed:
Address pre-commit suggestions

mitchnielsen committed:
Fix indentation on Deployment resource example

skyscrapr commented:
This initial PR adds support for deployments but does not support all attributes. 

However if you can merge this change first then deployment access resource PR can be opened while support for more attributes can be added in parallel. 

Let me know if that makes sense.

mitchnielsen committed:
Apply some fixes to the Deployment schema

Applies some fixes to the Deployment schema per the class:

https://github.com/PrefectHQ/nebula/blob/3836c49368641ffb97876d980cd90ada0168898c/src/prefect/server/schemas/core.py#L453

mitchnielsen commented:
> This initial PR adds support for deployments but does not support all attributes.
> 
> However if you can merge this change first then deployment access resource PR can be opened while support for more attributes can be added in parallel.
> 
> Let me know if that makes sense.

That seems fair to me. There are [quite a few attributes](https://github.com/PrefectHQ/nebula/blob/3836c49368641ffb97876d980cd90ada0168898c/src/prefect/server/schemas/core.py#L453) so I'm open to merging it with the essentials and we can add more in a follow-up PR.

mitchnielsen committed:
Remove 'status' field

The 'status' field isn't something the user will configure so let's just
remove it for now.

mitchnielsen committed:
Remove comments for un-implemented fields

We can add these back in if/when they're implemented.

mitchnielsen committed:
Re-enable commented fields in copy to model

These fields are already defined and supported so let's enable them.

mitchnielsen committed:
Remove IsScheduleActive (deprecated)

This field is being deprecated so let's not expose it for now.

https://github.com/PrefectHQ/prefect/pull/14794

mitchnielsen committed:
Remove document-id-related fields for now

Removes the StorageDocumentID and InfrastructureDocumentID fields for
now as they're not fully implemented yet.

mitchnielsen committed:
Consistent field formatting and implementation

Applies consistent field formatting and implementation.

mitchnielsen committed:
Fix Update method's comment and signature

- Fixes a copy/pasta error in the method comment
- Refers to the id rather than the workspaceID as expected

mitchnielsen committed:
Apply fixes to schema attribute properties

Applies fixes to the schema attribute properties regarding optionality,
requiredness, and planmodifiers

mitchnielsen committed:
Finish implementing ManifestPath

mitchnielsen committed:
Sort struct keys alphabetically

Lots of keys on these structs, let's alphabetize so it's easier to
parse them.

mitchnielsen committed:
Clean up import state functionality

mitchnielsen committed:
Add more fields to Deployment example

mitchnielsen committed:
Clean up Deployment tests

mitchnielsen committed:
Fix attributes on test deployment

mitchnielsen commented:
Getting close. Working on tests, and trying to figure out why updating any fields on the Deployment gives a `403`.

mitchnielsen committed:
Reference an existing work pool

mitchnielsen committed:
Use ephemeral workspace for deployment test

mitchnielsen committed:
Configure dependency on workspace and flow

mitchnielsen committed:
Disable Deployment Update test for now

mitchnielsen committed:
Fix arg to GetResourceWorkspaceImportStateID

mitchnielsen committed:
Fix order of accountid, workspaceid args

mitchnielsen committed:
Remove unsupported 'name' from DeploymentUpdate

mitchnielsen committed:
Re-enable Deployment update tests

mitchnielsen committed:
Use template for TF manifest, test all fields

Uses go's text/template package to make the manifest template easier
to parse visually. Using Sprintf with positional arguments was becoming
unwieldy with so many fields.

Ensures that all fields are tested. There are two exceptions which have
associated comments explaining why they're disabled.

mitchnielsen committed:
Use evergreen workspace for now

The ephemeral workspaces setup seemed to be causing 503 Service
Unavailable errors. It could be related to creating too many workspaces
at once. So for now, let's use the evergreen "github-ci-tests"
workspace.

mitchnielsen committed:
Test work pool and queue names

github-actions[bot] committed:
Generate Terraform Docs

mitchnielsen committed:
Rename example work pool name

mitchnielsen commented on docs/resources/flow.md:
````diff
@@ -50,5 +50,5 @@ Import is supported using the following syntax:
 ```shell
 # Prefect Flows can be imported via flow_id,workspace_id
-terraform import prefect_variable.example 00000000-0000-0000-0000-000000000000,00000000-0000-0000-0000-000000000000
+terraform import prefect_flow.example 00000000-0000-0000-0000-000000000000,00000000-0000-0000-0000-000000000000
````
This is just a drive-by fix I saw when working on testing Deployments with Flows.

parkedwards replied:
copy pasta strikes again

github-actions[bot] committed:
Generate Terraform Docs

mitchnielsen commented on internal/api/deployments.go:
```diff
@@ -0,0 +1,110 @@
+	WorkQueueName          string   `json:"work_queue_name,omitempty"`
+}
+
+// DeploymentFilter defines the search filter payload
+// when searching for deployements by name.
+// example request payload:
+// {"deployments": {"handle": {"any_": ["test"]}}}.
+type DeploymentFilter struct {
+}
```
There are some extra structs here and below that aren't implemented yet, but I didn't want to delete them and have to recreate them later. Happy to do that though if desired.

parkedwards replied:
happy to keep them in

mitchnielsen commented on internal/provider/helpers/template.go:
```diff
@@ -0,0 +1,28 @@
+	"text/template"
+)
+
+// RenderTemplate is a helper function to take a template and some configuration and
+// return the result as a string.
+//
+// Where:
+//   - tpl is a string like "Hello, {{.Name}}"
+//   - data is a struct like `mydata{Name: "World"}`
+func RenderTemplate(tpl string, data interface{}) string {
```
The Terraform manifest building got really unwieldy with all of the positional arguments to `Sprintf`, so I used Go's `text/template` here so it's a bit easier to read and maintain.

mitchnielsen committed:
Merge branch 'main' of github.com:PrefectHQ/terraform-provider-prefect into resource_deployment

parkedwards committed:
Merge branch 'main' into resource_deployment

parkedwards commented on examples/resources/prefect_deployment/resource.tf:
```diff
@@ -0,0 +1,27 @@
+}
+
+resource "prefect_deployment" "deployment" {
+  name                     = "%s"
```
```suggestion
  name                     = "my-deployment"
```

parkedwards commented on internal/client/deployments.go:
```diff
@@ -0,0 +1,187 @@
+	return deployments, nil
+}
+
+// Get returns details for a Workspace by ID.
```
```suggestion
// Get returns details for a Deployment by ID.
```

mitchnielsen committed:
Fix Get method comment

Co-authored-by: Edward Park <ed.sh.park@gmail.com>

mitchnielsen committed:
Fix deployment name in example

Co-authored-by: Edward Park <ed.sh.park@gmail.com>

mitchnielsen commented:
Thanks for getting this one off the ground @skyscrapr 🤝
